packages/flutter_test/lib/src/binding.dart

```diff
@@ -705,8 +705,10 @@ abstract class TestWidgetsFlutterBinding extends BindingBase
       // alone so that we don't cause more spurious errors.
       runApp(Container(key: UniqueKey(), child: _postTestMessage)); // Unmount any remaining widgets.
       await pump();
+      _pendingExceptionDetails = null;
```

Comment:
this is a change i am extremely uncomfortable with.

Because the globalkey reservation will not halt the tree build after my change. sometimes the tree will be left in a broken state if tests expect build exception. I think this is always the case, and my change just add one more possibility to it.

If the element tree is unstable, the runApp might fail with flutter error when it try to clean up unstable tree. Other than muting the exception, I can also create a method to clean up unstable tree. However, that might require expose some api from Element or RenderObject, which i am not sure if it worth it.

Reply:
We definitely don't want to swallow all errors here, because it would swallow errors in the dispose methods that we want to catch.

What's the exception you get here if you previously failed due to duplicate global keys?

Reply:
```'element._debugLifecycleState == _ElementLifecycle.active': is not true.```
Because the one of the test will make a element with global key coexist in two place in the tree
'GlobalKey duplication 15 - duplicating during build - before'

When the runapp in line 706 try to tear down the tree, it the element will gets deactivated twice. The second time it deactivates will throw this error.

Reply:
Can we tweak when we throw the exception so that it happens after deactivation while still catching the problems?

Reply:
In the _debugVerifyGlobalKeyReservation when we detect an illegal state, can we just "fix" it there by removing one of the duplicates? That would avoid the error caused by the runApp here.

(The whole point of `runApp(Container(key: UniqueKey(), child: _postTestMessage));` is to catch errors that may happen when the tree from the test is taken down. Swallowing all exceptions here is defeating that purpose.)

Reply:
I think it make sense to fix it so that the tree will not be broken. Another question is do we care if tree is not broken in the case of test that expect failure?

The other way to work around this will be detect if any exception has happened in flutter.onError, and then only attempt to tear down the tree when there is none

Reply:
Update: it seems like it is necessary to clean up the tree as long as the test has been finished without pending exception, since we will have to tear down the tree anyway in the next test.